### PR TITLE
Feat/cancel hearings

### PR DESCRIPTION
**Pull request by HarryH96**
### JIRA link (if applicable) ###

https://tools.hmcts.net/jira/browse/SSCS-10328
https://tools.hmcts.net/jira/browse/SSCS-10335
https://tools.hmcts.net/jira/browse/SSCS-10336
https://tools.hmcts.net/jira/browse/SSCS-10337
https://tools.hmcts.net/jira/browse/SSCS-10338
https://tools.hmcts.net/jira/browse/SSCS-10339
https://tools.hmcts.net/jira/browse/SSCS-10340
https://tools.hmcts.net/jira/browse/SSCS-10341
https://tools.hmcts.net/jira/browse/SSCS-10342
https://tools.hmcts.net/jira/browse/SSCS-10344
https://tools.hmcts.net/jira/browse/SSCS-10345
https://tools.hmcts.net/jira/browse/SSCS-10346
https://tools.hmcts.net/jira/browse/SSCS-10641

### Change description ###
- Combines all cancel hearing PRs

**Does this PR introduce a breaking change?** (check one with "x")

```
[ ] Yes
[x] No
```

### CODE REVIEW

File: src/main/java/uk/gov/hmcts/reform/sscs/ccd/presubmit/postponementrequest/PostponementRequestAboutToStartHandler.java [hunk 1]
```diff
@@ -37,6 +38,10 @@ public PreSubmitCallbackResponse<SscsCaseData> handle(CallbackType callbackType,
 
         final PreSubmitCallbackResponse<SscsCaseData> response = new PreSubmitCallbackResponse<>(sscsCaseData);
 
+        if (HearingRoute.GAPS.equals(sscsCaseData.getRegionalProcessingCenter().getHearingRoute())) {
```
Comment (bhssamray) [inline src/main/java/uk/gov/hmcts/reform/sscs/ccd/presubmit/postponementrequest/PostponementRequestAboutToStartHandler.java:41]: Which change is this relating to?

File: src/main/java/uk/gov/hmcts/reform/sscs/model/AppConstants.java [hunk 2]
```diff
@@ -35,11 +35,10 @@ public class AppConstants {
     public static final String DWP_DOCUMENT_EVIDENCE_FILENAME_PREFIX = "FTA evidence received";
     public static final String DWP_DOCUMENT_EDITED_RESPONSE_FILENAME_PREFIX = "FTA edited response received";
     public static final String DWP_DOCUMENT_EDITED_EVIDENCE_FILENAME_PREFIX = "FTA edited evidence received";
+    public static final String POSTPONEMENTS_NOT_POSSIBLE_FOR_GAPS = "Postponement requests cannot be made for hearings listed in GAPS";
```
Comment (bhssamray) [inline src/main/java/uk/gov/hmcts/reform/sscs/model/AppConstants.java:38]: This is not related to the cancels hearing ?

File: build.gradle [hunk 3]
```diff
@@ -222,7 +222,7 @@ def versions = [
         reformLogging   : '5.1.7',
         sonarPitest     : '0.5',
         springfoxSwagger: '2.8.0',
-        sscsCommon      : '4.9.8',
+        sscsCommon      : '4.9.7-QA-TEST-f098937',
```
Comment (bhssamray) [inline build.gradle:225]: We can't merge the sscsCommon version change as is. We'll have to make a separate common tag and use it as per the guidelines.

File: README.md [hunk 4]
```diff
@@ -4,6 +4,13 @@
 Tribunals case api is a spring boot based application to create new appeals for the SSCS Appellants
 
 
+
+
```
Comment (bhssamray) [inline README.md:8]: Mot sure about this emptylines

**Review by vinothon84 — APPROVED** [review]

File: src/main/java/uk/gov/hmcts/reform/sscs/ccd/presubmit/actionpostponementrequest/ActionPostponementRequestAboutToSubmitHandler.java [hunk 5]
```diff
@@ -75,6 +84,24 @@ public PreSubmitCallbackResponse<SscsCaseData> handle(CallbackType callbackType,
         return response;
     }
 
+    private void cancelHearing(SscsCaseData sscsCaseData) {
+        log.info("Action postponement request: Check eligible hearing request for case {}", sscsCaseData
```
Comment (dattipoeHMCTS) [inline src/main/java/uk/gov/hmcts/reform/sscs/ccd/presubmit/actionpostponementrequest/ActionPostponementRequestAboutToSubmitHandler.java:88]: I really don't think we need all this info logging

Comment (dattipoeHMCTS) [reply src/main/java/uk/gov/hmcts/reform/sscs/ccd/presubmit/actionpostponementrequest/ActionPostponementRequestAboutToSubmitHandler.java:88]: Especially if we might not even execute the cancellation

File: src/main/java/uk/gov/hmcts/reform/sscs/ccd/presubmit/deathofappellant/DeathOfAppellantAboutToSubmitHandler.java [hunk 6]
```diff
@@ -94,10 +105,30 @@ public PreSubmitCallbackResponse<SscsCaseData> handle(CallbackType callbackType,
...
             preSubmitCallbackResponse.getData().setIsConfidentialCase(null);
         }
+        cancelHearing(callback);
 
         return preSubmitCallbackResponse;
     }
 
+    private void cancelHearing(Callback<SscsCaseData> callback) {
+        SscsCaseData sscsCaseData = callback.getCaseDetails().getCaseData();
+        log.info("Death of appellant: Cancel hearing conditions ({}) ({}) ({}) for case ({})",
```
Comment (dattipoeHMCTS) [inline src/main/java/uk/gov/hmcts/reform/sscs/ccd/presubmit/deathofappellant/DeathOfAppellantAboutToSubmitHandler.java:115]: Same here

File: src/main/java/uk/gov/hmcts/reform/sscs/ccd/presubmit/issuefinaldecision/IssueFinalDecisionAboutToSubmitHandler.java [hunk 7]
```diff
@@ -67,26 +83,57 @@ public PreSubmitCallbackResponse<SscsCaseData> handle(CallbackType callbackType,
...
-            if (sscsCaseData.getSscsFinalDecisionCaseData().getWriteFinalDecisionPreviewDocument() != null) {
-                createFinalDecisionNoticeFromPreviewDraft(preSubmitCallbackResponse);
-                clearTransientFields(preSubmitCallbackResponse);
-            } else {
-                preSubmitCallbackResponse.addError("There is no Preview Draft Decision Notice on the case so decision cannot be issued");
-            }
+        sscsCaseData.setDwpState(FINAL_DECISION_ISSUED.getId());
+        createFinalDecisionNoticeFromPreviewDraft(preSubmitCallbackResponse);
+        clearTransientFields(preSubmitCallbackResponse);
+        log.info("Issue final decision request: Check eligible hearing request for case {}", sscsCaseData.getCcdCaseId());
```
Comment (dattipoeHMCTS) [inline src/main/java/uk/gov/hmcts/reform/sscs/ccd/presubmit/issuefinaldecision/IssueFinalDecisionAboutToSubmitHandler.java:95]: Whoa

File: src/main/java/uk/gov/hmcts/reform/sscs/ccd/presubmit/dormant/DormantEventsAboutToSubmitHandler.java [hunk 8]
```diff
@@ -46,6 +74,42 @@ public PreSubmitCallbackResponse<SscsCaseData> handle(CallbackType callbackType,
...
+
         return new PreSubmitCallbackResponse<>(caseData);
     }
+
+    private final Predicate<Callback<SscsCaseData>> eligibleForHearingsCancel = callback ->
+            SscsUtil.isValidCaseState(callback.getCaseDetailsBefore().map(CaseDetails::getState)
+                    .orElse(State.UNKNOWN), List.of(State.HEARING, State.READY_TO_LIST))
+            && SscsUtil.isSAndLCase(callback.getCaseDetails().getCaseData())
+            && isValidCaseEventForCancellation(callback.getEvent())
+            && isScheduleListingEnabled;
```
Comment (dattipoeHMCTS) [inline src/main/java/uk/gov/hmcts/reform/sscs/ccd/presubmit/dormant/DormantEventsAboutToSubmitHandler.java:110]: You could short circuit this by moving isScheduleListingEnabled to the front

File: src/main/java/uk/gov/hmcts/reform/sscs/ccd/presubmit/notlistable/NotListableAboutToSubmitHandler.java [hunk 9]
```diff
@@ -42,4 +65,10 @@ public PreSubmitCallbackResponse<SscsCaseData> handle(CallbackType callbackType,
     private void clearTransientFields(SscsCaseData sscsCaseData) {
         sscsCaseData.setNotListableDueDate(null);
     }
+
+    private final Predicate<Callback<SscsCaseData>> eligibleForHearingsCancel = callback ->
+            SscsUtil.isValidCaseState(callback.getCaseDetailsBefore().map(CaseDetails::getState)
+                    .orElse(State.UNKNOWN), List.of(State.HEARING, State.READY_TO_LIST))
+                    && SscsUtil.isSAndLCase(callback.getCaseDetails().getCaseData())
+                    && isScheduleListingEnabled;
```
Comment (dattipoeHMCTS) [inline src/main/java/uk/gov/hmcts/reform/sscs/ccd/presubmit/notlistable/NotListableAboutToSubmitHandler.java:73]: Same again

File: src/main/java/uk/gov/hmcts/reform/sscs/ccd/presubmit/struckout/StruckOutAboutToSubmitHandler.java [hunk 10]
```diff
@@ -35,7 +49,25 @@ public PreSubmitCallbackResponse<SscsCaseData> handle(CallbackType callbackType,
...
 
         sscsCaseData.setDwpState(DwpState.STRIKE_OUT_ACTIONED.getId());
-
+        cancelHearing(callback);
         return new PreSubmitCallbackResponse<>(sscsCaseData);
     }
+
+    private void cancelHearing(Callback<SscsCaseData> callback) {
+        SscsCaseData sscsCaseData = callback.getCaseDetails().getCaseData();
+        log.info("Strike out case: Cancel hearing conditions ({}) ({}) ({}) for case ({})", isScheduleListingEnabled,
```
Comment (dattipoeHMCTS) [inline src/main/java/uk/gov/hmcts/reform/sscs/ccd/presubmit/struckout/StruckOutAboutToSubmitHandler.java:58]: Again

File: src/main/java/uk/gov/hmcts/reform/sscs/ccd/presubmit/voidcase/VoidCaseAboutToSubmitHandler.java [hunk 11]
```diff
@@ -39,9 +55,29 @@ public PreSubmitCallbackResponse<SscsCaseData> handle(CallbackType callbackType,
...
+        cancelHearing(callback);
 
         PreSubmitCallbackResponse<SscsCaseData> callbackResponse = new PreSubmitCallbackResponse<>(sscsCaseData);
 
         return callbackResponse;
     }
+
+    private void cancelHearing(Callback<SscsCaseData> callback) {
+        SscsCaseData sscsCaseData = callback.getCaseDetails().getCaseData();
+        log.info("Void case: Cancel hearing conditions ({}) ({}) ({}) for case ({})", isScheduleListingEnabled,
```
Comment (dattipoeHMCTS) [inline src/main/java/uk/gov/hmcts/reform/sscs/ccd/presubmit/voidcase/VoidCaseAboutToSubmitHandler.java:67]: Same here

File: src/main/java/uk/gov/hmcts/reform/sscs/ccd/presubmit/withdrawnappeals/AdminAppealWithdrawnHandler.java [hunk 12]
```diff
@@ -41,10 +60,29 @@ public PreSubmitCallbackResponse<SscsCaseData> handle(CallbackType callbackType,
         if (nonNull(documentDetails)) {
             addToSscsDocuments(caseData, documentDetails);
         }
-
+        cancelHearing(callback);
         return new PreSubmitCallbackResponse<>(caseData);
     }
 
+    private void cancelHearing(Callback<SscsCaseData> callback) {
+        SscsCaseData sscsCaseData = callback.getCaseDetails().getCaseData();
```
Comment (dattipoeHMCTS) [inline src/main/java/uk/gov/hmcts/reform/sscs/ccd/presubmit/withdrawnappeals/AdminAppealWithdrawnHandler.java:68]: Same here

**Review by dattipoeHMCTS — COMMENTED** [review]
I really think those log messages should be trimmed down and if any are required, set to debug

File: src/test/java/uk/gov/hmcts/reform/sscs/ccd/presubmit/withdrawnappeals/AdminAppealWithdrawnHandlerTest.java [hunk 13]
```diff
@@ -2,25 +2,38 @@
...
+import uk.gov.hmcts.reform.sscs.ccd.presubmit.resendtogaps.ListAssistHearingMessageHelper;
 
 @RunWith(JUnitParamsRunner.class)
 public class AdminAppealWithdrawnHandlerTest extends AdminAppealWithdrawnBase {
+
+    @Rule
+    public MockitoRule rule = MockitoJUnit.rule();
+
+    @Mock
+    private ListAssistHearingMessageHelper hearingMessageHelper;
```
Comment (dattipoeHMCTS) [inline src/test/java/uk/gov/hmcts/reform/sscs/ccd/presubmit/withdrawnappeals/AdminAppealWithdrawnHandlerTest.java:32]: Are we missing a positive test case for this class?

**Review by vinothon84 — APPROVED** [review]

**Review by johnoneill46 — APPROVED** [review]
